libretroshare/src/retroshare/rsids.h

```diff
@@ -19,161 +20,219 @@
-// 			cached string convertion is not thread safe.
-//
-
 #pragma once
 
 #include <stdexcept>
 #include <string>
 #include <iostream>
 #include <ostream>
 #include <string.h>
-#include <stdint.h>
-#include <util/rsrandom.h>
+#include <cstdint>
 #include <vector>
 #include <list>
 #include <set>
+#include <memory>
 
-template<uint32_t ID_SIZE_IN_BYTES,bool UPPER_CASE,uint32_t UNIQUE_IDENTIFIER> class t_RsGenericIdType
-{
-	public:
+#include "util/rsdebug.h"
+#include "util/rsrandom.h"
+#include "util/stacktrace.h"
 
-		typedef std::list<t_RsGenericIdType<ID_SIZE_IN_BYTES,UPPER_CASE,UNIQUE_IDENTIFIER> > std_list;
-		typedef std::vector<t_RsGenericIdType<ID_SIZE_IN_BYTES,UPPER_CASE,UNIQUE_IDENTIFIER> > std_vector;
-		typedef std::set<t_RsGenericIdType<ID_SIZE_IN_BYTES,UPPER_CASE,UNIQUE_IDENTIFIER> > std_set;
+/**
+ * RsGenericIdType values might be random, but must be different, in order to
+ * make the various IDs incompatible with each other.
+ */
+enum class RsGenericIdType
+{
+	SSL,
+	PGP_ID,
+	SHA1,
+	PGP_FINGERPRINT,
+	GXS_GROUP,
+	GXS_ID,
+	GXS_MSG,
+	GXS_CIRCLE,
+	GROUTER,
+	GXS_TUNNEL,
+	DISTANT_CHAT,
+	NODE_GROUP,
+	SHA256,
+	BIAS_20_BYTES
+};
 
-		t_RsGenericIdType() 
-		{ 
-			memset(bytes,0,ID_SIZE_IN_BYTES) ; 	// by default, ids are set to null()
-		}
-		virtual ~t_RsGenericIdType() {}
-
-		// Explicit constructor from a hexadecimal string
-		//
-		explicit t_RsGenericIdType(const std::string& hex_string) ;
-
-		// Explicit constructor from a byte array. The array should have size at least ID_SIZE_IN_BYTES
-		//
-		explicit t_RsGenericIdType(const unsigned char bytes[]) ;
-
-		// Explicit constructor from a different type, checking that the sizes are compatible.
-		// This is used for conversions such as 
-		//
-		// 		GroupId -> CircleId
-		// 		GroupId -> GxsId
-		//
-		template<bool UPPER_CASE2,uint32_t UNIQUE_IDENTIFIER2>
-		explicit t_RsGenericIdType(const t_RsGenericIdType<ID_SIZE_IN_BYTES,UPPER_CASE2,UNIQUE_IDENTIFIER2>& id)
+/**
+ * This class aims at defining a generic ID type that is a list of bytes. It
+ * can be converted into a hexadecial string for printing, mainly) or for
+ * compatibility with old methods.
+ *
+ * To use this class, derive your own ID type from it.
+ * @see RsPpgFingerprint as an example.
+ *
+ * Take care to define and use a different @see RsGenericIdType for each ne type
+ * of ID you create, to avoid implicit conversion between subtypes, and
+ * therefore accidental ID mixup is impossible.
+ *
+ * ID Types with different lengths are not convertible even using explicit
+ * constructor and compilation would fail if that is attempted.
+ *
+ * Warning: never store references to a t_RsGenericIdType accross threads, since
+ * the cached string convertion is not thread safe.
+ */
+template<uint32_t ID_SIZE_IN_BYTES, bool UPPER_CASE, RsGenericIdType UNIQUE_IDENTIFIER>
+struct t_RsGenericIdType
+{
+	using Id_t = t_RsGenericIdType<ID_SIZE_IN_BYTES,UPPER_CASE,UNIQUE_IDENTIFIER>;
```

Comment:
sometimes you just rewrite for the pleasure of rewriting...and at the expense of me spending more time to read .

Reply:
this line improve the readability of the following code a lot, without all those template huge line around that make it difficult to read in my little laptop